cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll

```diff
@@ -0,0 +1,270 @@
+  }
+}
+
+class IteratorAddOperator extends Operator, TaintFunction {
+  IteratorAddOperator() {
+    this.hasName("operator+") and
+    (
+      this
+          .getACallToThisFunction()
+          .getArgument(0)
+          .getFullyConverted()
+          .getUnspecifiedType()
+          .(PointerType)
+          .getBaseType() instanceof LegacyIterator or
+      this
+          .getACallToThisFunction()
+          .getArgument(0)
+          .getFullyConverted()
+          .getUnspecifiedType()
+          .(PointerType)
+          .getBaseType() instanceof LegacyIterator
+    )
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    (
+      input.isParameter(0) or
+      input.isParameter(1)
+    ) and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * A non-member `operator-` function that takes an iterator as its first argument.
+ */
+class IteratorSubOperator extends Operator, TaintFunction {
+  IteratorSubOperator() {
+    this.hasName("operator-") and
+    this
+        .getACallToThisFunction()
+        .getArgument(0)
+        .getFullyConverted()
+        .getUnspecifiedType()
+        .(PointerType)
+        .getBaseType() instanceof LegacyIterator
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    (
+      input.isParameter(0) or
+      input.isParameter(1)
+    ) and
+    output.isReturnValue()
+  }
+}
+
+class IteratorAssignArithmeticOperator extends MemberFunction, DataFlowFunction, TaintFunction {
+  IteratorAssignArithmeticOperator() {
+    (
+      this.hasName("operator+=") or
+      this.hasName("operator-=")
+    ) and
+    this.getDeclaringType() instanceof LegacyIterator
+  }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameter(0) and
+    output.isReturnValue()
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameterDeref(1) and
+    output.isParameterDeref(0)
+  }
+}
+
+class IteratorPointerDereferenceMemberOperator extends MemberFunction, TaintFunction {
+  IteratorPointerDereferenceMemberOperator() {
+    this.hasName("operator*") and
+    this.getDeclaringType() instanceof LegacyIterator
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isQualifierObject() and
+    output.isReturnValue()
+  }
+}
+
+class IteratorCrementMemberOperator extends MemberFunction, DataFlowFunction, TaintFunction {
+  IteratorCrementMemberOperator() {
+    (
+      this.hasName("operator++") or
+      this.hasName("operator--")
+    ) and
+    this.getDeclaringType() instanceof LegacyIterator
+  }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    input.isQualifierAddress() and
+    output.isReturnValue()
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isQualifierObject() and
+    output.isQualifierObject()
```

Comment:
I think this is assumed for member function calls?  (I actually need to block this flow for `clear` methods)